### PR TITLE
FilePathLayout - Faster rendering of filename, when not ThreadAgnostic

### PR DESCRIPTION
**Pull request by snakefoot**
#1071 improved from 550.000 msgs/sec to 1.500.000 msgs/sec.

---
This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1932)

### CODE REVIEW

File: src/NLog/Internal/FilePathLayout.cs [hunk 1]
```diff
@@ -164,6 +164,13 @@ private string GetRenderedFileName(LogEventInfo logEvent, System.Text.StringBuil
 
             if (reusableBuilder != null)
             {
+                if (!_layout.IsThreadAgnostic)
```
Comment (304NotModified) [inline src/NLog/Internal/FilePathLayout.cs]: I'm doubting if this should be `if (_layout.IsThreadAgnostic)`, as ThreadAgnostic stuff could be calculated on every thread. (and thus could be lazy). 

non-ThreadAgnostic should be calculated on the main thread.

Comment (snakefoot) [inline src/NLog/Internal/FilePathLayout.cs]: IsThreadAgnostic means that it doesn't have to be precalculated by async-wrappers or buffering-wrappers. As none of the layouts requires the context-information of the logging-thread.

But it it requires the context-information of the logging-thread, and there is a async-wrapper or buffering-wrapper. Then it is possible to extract the precalculated layout-rendering-result.

Comment (snakefoot) [inline src/NLog/Internal/FilePathLayout.cs]: It actually checks if there is a chance that the main-thread has already rendered the logevent using the layout. And if so then dives into the logevent-cache of pre-rendered-layout-results.

**Comment by 304NotModified** [comment]
Thanks for the explanation!

**Pull request merged**